### ISSUE TITLE
Should the coding standard mention PEP8

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10821]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)| @puzzlement reported|
|-|-|
|Trac ID|trac#1220|
|Type|defect|
|Created|2005-09-28 07:33:04Z|

Attachments:

* [1220.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/717/7173e41e4de7a780906fada68a682817c8803ed2/b9c807d0ad041cac093a904d18f2436df803d0aa.diff) (856 bytes) - added by therve on 2007-09-28 08:43:11Z - 
* [1220_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/717/7173e41e4de7a780906fada68a682817c8803ed2/93dd137d9f113647c501f8f09bbe1f52dc79f9d0.diff) (2825 bytes) - added by therve on 2007-10-01 10:39:20Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1220 1220
type__defect defect
reporter__hypatia hypatia
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1127892784000000 1127892784000000
changetime__1193129272000000 1193129272000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__spiv cc__itamarst cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Adding a lot of opinionated people to the nosy list on this one, since
twisted-bugs is broken. Feel free to remove yourselves if you are opinion-free.

At least a few people think that Twisted code needs to be PEP 8 style, as per
this conversation:

&lt;hypatia> does twisted follow pep 8?
&lt;mumak> hypatia: not all the time. mostly though
&lt;mumak> or maybe I'm mistaken.
&lt;hypatia> It's not in the coding standard
&lt;mumak> no.
&lt;lifeless> hypatia: during the sprint, many pep8 garhs were seen and corrected -
not by me
&lt;lifeless> hypatia: so at least some folk do ;0
&lt;hypatia> lifeless: by whom then?
&lt;lifeless> radix

(lifeless's "not me" thing may be inspired by working on or at least near
launchpad.net code, which seems to turn people into raving PEP 8 folks. spiv's
been known to reformat my code PEP 8 even when that means that it would no
longer actually fit the target standard. ;) )

So, since I hate undocumented "everyone knows" policy, does Twisted code need to
be PEP8 or is our code purely subject to the coding standard as per:

&lt;radix> hrm. Twisted doesn't follow PEP 8. The coding standard is
self-contained, and is close to a superset of PEP 8.

It seems like if that's the official line, it should go in the coding standard,
since people are starting to reference PEP 8 when criticising each other's code.
Comments?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
If it's not in trunk/doc/core/howto/policy/coding-standard.xhtml, it's not real.
 It's entirely possible there are things which should be made real, though.

Specific suggestions?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
spiv might be the person to give an overview of whether it's worth being PEP8
levels of cranky...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I attached a simple patch mentioning PEP 8. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

[#2101](https://github.com/twisted/twisted/issues/2101) was closed as a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Some conflicts which exist but which aren't actually documented in coding-standard.xhtml:

  * no backslash line continuations allowed
  * separate top-level classes and functions with three blank lines, class-level functions with two blank lines, don't ever omit the blank lines, don't ever use ^L
  * ASCII only (if we ever allowed non-ascii, we will allow UTF-8, not latin-1)
  * Use "as" syntax to resolve naming conflicts between different imports, not fully-qualified names (last clause in PEP8 "Imports" section)
  * The opening """ of a docstring should be on a line by itself

Some parts of coding-standard.xhtml are just out of date and wrong:
  * No "placeholder docstring" junk allowed - write as complete and descriptive a docstring as you can.  If it changes later, then it changes later, no big deal.
  * The single quote/emacs thing isn't relevant anymore
  * The |-prefix thing is wrong, code examples should be indented per epytext rules.
  * We don't update ChangeLog anymore, the release manager does NEWS at release time
  * In the "Methods" section, the example of how to add multiple callbacks is terrible and wrong.

The first set of these should probably be fixed before we add the PEP8 fallback.  The second set can be addressed separately.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Thanks for the review, I've addressed the first points in the last patch. The second set should be done in [#2217](https://github.com/twisted/twisted/issues/2217).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml set owner to @therve|
|-|-|

 * Regarding long lines, we cannot use parentheses to wrap long lines in import statements while we support Python 2.3. I believe our standard is to repeat the import, e.g.

```python
from very.long.package import foo, bar, baz
from very.long.package import qux, quux, quuux
```

 * The second diff hunk says "musn't", change to "must not".

Once these are changed, please land this branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [21396]) Make the coding standard mention PEP8 as fallback, and resolve the conflicts
between the two.

Author: therve
Reviewers: jml, exarkun
Fixes [#10821](https://github.com/twisted/twisted/issues/10821)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
